### PR TITLE
enhancement(http_server and heroku_logs sources): Added wildcard support for query parameters

### LINKED ISSUE #21321: Get all query parameters from http server source
# Use cases
Hi everyone.
We're using vector to aggregate HTTP requests. Currently, query parameters must be specified, by name, or they won't be outputted.

There's no easy way to support requests with varying arguments.

# References
Discussion on a similar issue #16396 

# Solution
I've updated the code to output all query parameters, as a string, to field `query_parameters_string`.
A configuration variable, `all_qp`, enables this functionality - It's off by default.

Here's an example configuration:
```yaml
  sources:
    http:
      type: http_server
      address: 127.0.0.1:8080
      encoding: json
      headers:
        - "*"
      method: POST
      path: /mypath/
      response_code: 200
      all_qp: true # Toggle on

  transforms:
    enriched_http:
      type: remap
      inputs:
        - http
      source: |-
        .params = parse_query_string!(.query_parameters_string)
  sinks:
    # Debug http
    stdout:
      type: console
      inputs: [enriched_http]
      encoding:
        codec: json    
```

# Proposal
I'd like to contribute this change back to the upstream project.
If you approve opening a PR I'd appreciate suggestions regarding functionality and testing.

# Notes
The Heroku Logplex and http server sources reuse functions so this change will add the capability to both.

### PR DESCRIPTION
**Pull request by uricorin**
Added wildcard support for query parameters in http_server and heroku_logs sources.
This resolves #21321

Example:
```yaml
  sources:
    all_query_parameters:
      type: http_server
      headers:
        - "*"
      query_parameters:
        - "*"
    param_glob:
      type: http_server
      headers:
        - "*"
      query_parameters:
        - "param*"
        - "exact_parameter"
```

### CODE REVIEW

File: website/cue/reference/components/sources/base/heroku_logs.cue [hunk 1]
```diff
@@ -439,12 +439,16 @@ base: components: sources: heroku_logs: configuration: {
 		description: """
 			A list of URL query parameters to include in the log event.
 
+			Accepts the wildcard (`*`) character for query parameters matching a specified pattern.
+
+			Specifying "*" results in all query parameters included in the log event.
```
Comment (neko-dd) [inline website/cue/reference/components/sources/base/heroku_logs.cue:444]: I'm not 100% sure what this means - can you elaborate?

Comment (uricorin) [reply website/cue/reference/components/sources/base/heroku_logs.cue:444]: Given the following query parameters: `hello=world&source=test&application=demo`:
##### Exact
```yaml
  sources:
    exact:
      type: http_server
      headers:
        - "*"
      query_parameters:
        - "hello"
```
Will emit `"message": {"query_parameters": {"hello": "world"}}` as part of the event

##### Wildcard
```yaml
  sources:
    exact:
      type: http_server
      headers:
        - "*"
      query_parameters:
        - "*"
```
Will emit `"message": {"query_parameters": {"hello": "world", "source": "test", "application": "demo"}}` as part of the event

##### Prefix
```yaml
  sources:
    exact:
      type: http_server
      headers:
        - "*"
      query_parameters:
        - "s*"
```
Will emit `"message": {"query_parameters": {"source": "test"}}` as part of the event

File: src/sources/util/http/query.rs [hunk 2]
```diff
@@ -6,26 +6,63 @@ use vector_lib::{
...
-                    log,
-                    Some(LegacyKey::Overwrite(path!(query_parameter_name))),
-                    path!("query_parameters", query_parameter_name),
-                    crate::event::Value::from(value.map(String::to_owned)),
-                );
+    let query_parameters_config = build_param_matcher(&remove_duplicates(
+        query_parameters_config.to_vec(),
+        "query_parameters",
+    ));
+    for qp in &query_parameters_config.unwrap() {
```
Comment (jszwedko) [inline src/sources/util/http/query.rs:22]: Why is this `unwrap` safe?

**Review by jszwedko — COMMENTED** [review]
Nice, thanks @uricorin ! I left a question inline, but other than that do you mind adding a changelog fragment? See https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md

File: src/sources/util/http/query.rs (hunk 2 above)
Comment (uricorin) [reply src/sources/util/http/query.rs]: It's a mistake on my part. I misunderstood the nuances between unwrap and '?'.
As far as I can tell, some patterns **would** result in errors.

I've moved the glob parsing to fn `build` (same as `headers`) so they'd fail in a safe manner.

Comment (jszwedko) [reply src/sources/util/http/query.rs]: Thanks! That makes sense to "compile" the patterns at start up to propagate any errors.

**Pull request merged**